<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Setting up a phylochemistry analysis | Phylochemistry</title>
  <meta name="description" content="Vingette for the R Package Phylochemistry." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Setting up a phylochemistry analysis | Phylochemistry" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Vingette for the R Package Phylochemistry." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Setting up a phylochemistry analysis | Phylochemistry" />
  
  <meta name="twitter:description" content="Vingette for the R Package Phylochemistry." />
  

<meta name="author" content="Lucas Busta" />


<meta name="date" content="2020-03-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="analyzing-chromatography-data-with-phylochemistry.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.0.1" data-path="index.html"><a href="index.html#install-phylochemistry-directly-from-github-and-view-documentation-contents"><i class="fa fa-check"></i><b>1.0.1</b> Install <code>phylochemistry</code> directly from github and view documentation contents</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html"><i class="fa fa-check"></i><b>2</b> Setting up a <code>phylochemistry</code> analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#conventions-in-phylochemsitry"><i class="fa fa-check"></i><b>2.1</b> Conventions in <code>phylochemsitry</code></a><ul>
<li class="chapter" data-level="2.1.1" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#monolists"><i class="fa fa-check"></i><b>2.1.1</b> Monolists</a></li>
<li class="chapter" data-level="2.1.2" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#storing-your-data-as-a-list-of-lists"><i class="fa fa-check"></i><b>2.1.2</b> Storing your data as a list of lists</a></li>
<li class="chapter" data-level="2.1.3" data-path="setting-up-a-phylochemistry-analysis.html"><a href="setting-up-a-phylochemistry-analysis.html#nomenclature"><i class="fa fa-check"></i><b>2.1.3</b> Nomenclature</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html"><i class="fa fa-check"></i><b>3</b> Analyzing chromatography data with <code>phylochemistry</code></a><ul>
<li class="chapter" data-level="3.1" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html#preparing-the-raw-data-and-metadata"><i class="fa fa-check"></i><b>3.1</b> Preparing the raw data and metadata</a></li>
<li class="chapter" data-level="3.2" data-path="analyzing-chromatography-data-with-phylochemistry.html"><a href="analyzing-chromatography-data-with-phylochemistry.html#using-the-phylochemistry-integrationapp"><i class="fa fa-check"></i><b>3.2</b> Using the <code>phylochemistry</code> integrationApp</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="gene-mining.html"><a href="gene-mining.html"><i class="fa fa-check"></i><b>4</b> Gene mining</a><ul>
<li class="chapter" data-level="4.1" data-path="gene-mining.html"><a href="gene-mining.html#blast"><i class="fa fa-check"></i><b>4.1</b> BLAST</a><ul>
<li class="chapter" data-level="4.1.1" data-path="gene-mining.html"><a href="gene-mining.html#using-the-blasttranscriptomes-command"><i class="fa fa-check"></i><b>4.1.1</b> Using the blastTranscriptomes command</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Phylochemistry</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="setting-up-a-phylochemistry-analysis" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Setting up a <code>phylochemistry</code> analysis</h1>
<p><code>phylochemistry</code> is not just a set of functions that helps with analyses. It is also a conceptual and organizational framework that helps you perform analyses very efficiently and with great versatility. It takes a few minutes to understand and set up a phylochemistry analysis, but that investment will be paid back many-fold over the course of a project’s lifetime.</p>
<div id="conventions-in-phylochemsitry" class="section level2">
<h2><span class="header-section-number">2.1</span> Conventions in <code>phylochemsitry</code></h2>
<div id="monolists" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Monolists</h3>
<p>One way to think of <code>phylochemistry</code> is as a set of tools that enables you to integrate and analyze raw data stored on your hard drive. Let’s use the example dataset included with <code>phylochemistry</code> to explore this idea. First, let’s load the phylochemistry library and set up our phylochemistry directory:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(phylochemistry)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">phylochemistry_example_dir &lt;-<span class="st"> &quot;/Users/lucasbusta/Desktop/phylochemistry_example_data/&quot;</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">phylochemistry_example_dir</a></code></pre></div>
<pre><code>## [1] &quot;/Users/lucasbusta/Desktop/phylochemistry_example_data/&quot;</code></pre>
<p>Now, let’s see what’s in that directory</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">dir</span>(phylochemistry_example_dir)</a></code></pre></div>
<pre><code>## [1] &quot;chem_data&quot;             &quot;monolists&quot;             &quot;sequences_by_species&quot; 
## [4] &quot;sequences_of_interest&quot;</code></pre>
<p>We can see that there’s a folder called monolists. What’s in there?</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">dir</span>(<span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;monolists/&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;blast_hits.csv&quot;        &quot;chemistry_samples.csv&quot;</code></pre>
</div>
<div id="storing-your-data-as-a-list-of-lists" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Storing your data as a list of lists</h3>
<p>There are lots of different kinds of data structures. <code>phylochemistry</code> was designed with a very simple and flexible data structure in mind: a heirarchy of lists. These are very easy to create and modify:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">my_analysis &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">my_analysis<span class="op">$</span>chemistry_data &lt;-<span class="st"> </span><span class="kw">readMonolist</span>(<span class="dt">monolist_in_path =</span> <span class="kw">paste0</span>(phylochemistry_example_dir, <span class="st">&quot;monolists/chemistry_samples.csv&quot;</span>))</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">my_analysis<span class="op">$</span>chemistry_data</a></code></pre></div>
<pre><code>## # A tibble: 9 x 4
##   sample_name amount_internal_std_in_… amount_mixed_analyte_std_… replicate
##   &lt;chr&gt;                          &lt;int&gt;                      &lt;int&gt;     &lt;int&gt;
## 1 10_10_1                           10                         10         1
## 2 10_10_2                           10                         10         2
## 3 10_10_3                           10                         10         3
## 4 10_20_1                           10                         20         1
## 5 10_20_2                           10                         20         2
## 6 10_20_3                           10                         20         3
## 7 10_30_1                           10                         30         1
## 8 10_30_2                           10                         30         2
## 9 10_30_3                           10                         30         3</code></pre>
<p>This is a list of chemistry samples and how much of two different types of standards are in each.</p>
</div>
<div id="nomenclature" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Nomenclature</h3>
<!-- snakeCase -->
<!-- ## Setting up a new `phylochemistry` analysis -->
<!-- If you want, you can use the `setupPhylochemistryAnalysis` function to help you set up a new analysis. Otherwise, you can do it manually. Simply create a directory that contains subdirectories called "alignments", "chem_data", monolists" ... -->
<!-- `phylochemistry` enables analyses of natural chemicals and gene sequences in the context of phylogeny. -->
<!-- ### Install `phylochemistry` directly from github and view documentation contents -->
<!-- ```{r} -->
<!-- devtools::install_github("LucasBusta/phylochemistry", quiet = TRUE) -->
<!-- packages <- c("phylochemistry") -->
<!-- lapply(packages, require, character.only = TRUE) -->
<!-- help(package = "phylochemistry") -->
<!-- ``` -->
<!-- ## Visualizing data types separately -->
<!-- 1. $Taxa - taxonomy information   -->
<!--   + $species - we can see what's in the list just using `head()`. -->
<!-- ```{r} -->
<!-- data(fads) -->
<!-- head(fads$Taxa$species) -->
<!-- ``` -->
<!--   + $tree - we can see the tree using (among others) the excellent `ggtree` package -->
<!-- ```{r} -->
<!-- library(ggtree, quietly = TRUE) -->
<!-- ggtree(fads$Taxa$tree) -->
<!-- ``` -->
<!-- 2. $Chemistry - structural and quantitative chemical data   -->
<!--   + $structural_chemical_data -  -->
<!-- ## Building a phylochem object from scratch -->
<!-- ### Create a phylochem object called "fads" and bind taxonomic data to it -->
<!-- The data for this section is included in the "example_data" folder -->
<!-- ```{r} -->
<!-- species <- "/Users/lucasbusta/Documents/Science/_Lab_Notebook/_Phylochemistry/__plant_list_and_DBs/_phylochemistry_lists/angiosperms.csv" -->
<!-- tree <- "/Users/lucasbusta/Documents/Science/_Lab_Notebook/_Phylochemistry/__plant_list_and_DBs/PhytoPhylo.newick" -->
<!-- fads <- list() -->
<!-- fads <- loadTaxa(file = species, phylochem = fads, tree = tree) -->
<!-- ``` -->
<!-- ### Create a phylochem object called "fads" and bind taxonomic data to it -->
<!-- Let's have a look at the structure of the "fads" phylochem object so far. We can see that it contains two things: a data.frame ("species") with information from the "angiosperms" species file, and a phylogenetic tree ("tree"; actually a list), that contains an angiosperm phylogeny with ~31,000 species. -->
<!-- ```{r} -->
<!-- str(fads, strict.width = "cut") -->
<!-- ``` -->
<!-- ### Bind chemical data to the "fads" phylochem object -->
<!-- ```{r} -->
<!-- chemistry <- "/Users/lucasbusta/Documents/Science/_Lab_Notebook/__R_code/packages/phylochemistry/example_data/seed_oil_chemistry.csv" -->
<!-- fads <- loadChemistry(file = chemistry, phylochem = fads, structural_alignment = FALSE, fingerprint_alignment = FALSE) -->
<!-- str(fads, strict.width = "cut") -->
<!-- ``` -->

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="analyzing-chromatography-data-with-phylochemistry.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

</body>

</html>
